### PR TITLE
🏗️ Structure refactor T0 (discussion 31)

### PR DESCRIPTION
**Pull request by LeiRoF**
[Discussion 31](https://github.com/Gunivers/Gipsy/discussions/31)

ToDo List:
- [x] Plugin main file
- [x] Doc grouping
- [x] Config grouping
- [ ] Option (enabled by default) to install dependencies in a venv
- [ ] Move plugin database model
- [x] Setup.py
- [ ] I18N public singleton (no more in a cog)
- [ ] Sconfig public singleton (no more in a cog)
- [ ] Logs overhaul
- [ ] Start.py simplification

### CODE REVIEW

**Comment by Aeris1One** [comment]
LGTM, je pense qu'il serait judicieux de tester avec seulement l'Action bêta avant de d'activer l'autodeploy aussi sur master

**Comment by ascpial** [comment]
You should add the LRFutils module to the `requirements.txt` file

**Comment by ascpial** [comment]
I don't think storing the configuration in a python file is a good idea.

From a code perspective, a python file is not ideal: importing the file to read the configuration will miss default value if the user decide to empty the file; there is also security concerns: if a user give his configuration file to another user, he can insert code that will be executed in the workspace of the second user, generally runing code in the configuration is a bad idea. I should also mention that a python file cannot really handle configuration levels, like what you can do in a json or a ini file. Storing configurations for all plugins in one file could also be a bad idea, because the constant used by a plugin could have the same name as a constant from an other plugin, resoulting in conflicts (this type of issue can easely be resolved in a json or yaml file by requiring a namespace for each plugin).

If you create a module that handles default values, I don't see the advantages of using a python file instead of a json, yaml, ini file or whatever format you want.

**Comment by LeiRoF** [comment]
Using python as config files doesn't come from nowhere. Sphinx is using this method and it's proven to work.

### The problem you mentioned

> importing the file to read the configuration will miss default value if the user decide to empty the file

Ok, the user can empty everything without consideration of what he will break... the config using python is not the problem here.

> there is also security concerns: if a user give his configuration file to another user, he can insert code that will be executed in the workspace of the second user, generally runing code in the configuration is a bad idea

Yes, the user can accept a configuration file that will run weird code... as he can accept a plugin that will run weird code, the user is supposed to look at what he put in his bot... whatever the type of config you use, as long as the user put something in his bot without verification, it's dangerous.

> I should also mention that a python file cannot really handle configuration levels, like what you can do in a json or a ini file.

wdym? As a JSON file is strictely equivalent to a python dictionary, nothing can be done using JSON that cannot using python 🤔

> Storing configurations for all plugins in one file could also be a bad idea, because the constant used by a plugin could have the same name as a constant from an other plugin, resoulting in conflicts (this type of issue can easely be resolved in a json or yaml file by requiring a namespace for each plugin).

Then it's more a question of conventions, it's also possible to apply the same type of convention using python files (ex: start every config variable with the name of the plugin, or put them in a class that have the name of the plugin ... or even using a unique dictionary for each plugin ! There is plenty of possible solutions). If two plugin share the same name (ie namespace), the problem is the same using json files.

### The advantage of this solution

Using python file allow to deal with variables instead of dictionnary entries, which is way more conveniant. Moreover, it's easier to enderstand that "import config" will... import the configuration variables contained in the file than using a function defined somewhere and that return something... that we have to check the doc to deal with.

Moreover, running code in the configuration can have benefits. It's a file called at the very begining of the bot start and it is never overwritten, so it allow for users to add pre-run procedure to the bot, such I did in docs/conf.py (run by sphinx)

### TL;DR

The bot is made to be user friendly, not to prevent user stupidity 
Python files are more permissive, it mean that we can do a lot of things with it... including bad things. But the setup tool, the comments (which are not possible in JSON) etc. allow users to only edit what they understand

**Comment by Aeris1One** [comment]
> Then it's more a question of conventions, it's also possible to apply the same type of convention using python files (ex: start every config variable with the name of the plugin, or put them in a class that have the name of the plugin ... or even using a unique dictionary for each plugin ! There is plenty of possible solutions). If two plugin share the same name (ie namespace), the problem is the same using json files.

Hum, actually, about conventions, maybe it would be better to enforce them, or else you'll be taking the risk of them not being respected. I would advocate restricting plugins to only seeing their own configuration, that would prevent a plugin trying to reinvent the wheel by reading itself some config which would otherwise be handled by the core.

> Using python file allow to deal with variables instead of dictionnary entries, which is way more conveniant. Moreover, it's easier to enderstand that "import config" will... import the configuration variables contained in the file than using a function defined somewhere and that return something... that we have to check the doc to deal with.

Well, we can make a config.py which setup all those variables by reading the config. We would still need to `import config` and that would import the exact same variables, excepted that they would have been extracted from a JSON/YAML file. That way it would also be possible to dynamically edit this config file. 
For example, upon plugin installation, some questions are asked to the user regarding the configuration to make and a namespace is added at the end of the configuration file.

e.g. for the friendly-ban plugin, upon installation you would be asked some questions like "which server would you like to enable friendly-ban in?" and then this may be added at the end of the config file:
```yaml
- friendly-ban:
  - enabled-servers:
    - 125723125685026816
    - 835218602511958116
  - disabled-events:
    # this is empty by default, feel free to add any event id you want to disable
```

Actually, I'm not sure about the `import config` method which would import all config vars, maybe it would be better to import instead some functions (like `config.get(enabled-servers)` which would return a list of children, here they're servers IDs) and restrict a plugin to only seeing its config option. Zero-trust policy.

**Comment by LeiRoF** [comment]
Ok, tell me what you think about that:

The plugin `config.py` file content:
image

When this file is squashed in the main `config.py` file, it gives this:
image

In that way, we avoid name conflits by always using a class that have the name of the plugin (there is still a possible conflict in plugin names which will be tedious to remove)

Also, as the `config.py` file is always run when the bot start (and also during the setup), the plugin can add some configuration steps.

**Comment by ascpial** [comment]
> Ok, the user can empty everything without consideration of what he will break... the config using python is not the problem here.

If the config is setup properly, there should be default values for settings that the user did not fill. This is not possible using the python file for the configuration directly, but if you setup a json-based system for example, the configuration reader could fill the empty values.

> Yes, the user can accept a configuration file that will run weird code... as he can accept a plugin that will run weird code, the user is supposed to look at what he put in his bot...

As Aeris1One said, we should be in a zero-trust policy: the user is more likely to edit the configuration file that install plugins that are not ship with the installer (I think we will make an installer to improve the user accessibility).

> nothing can be done using JSON that cannot using python

It's much more intuitive to store nested data in JSON than in python.
I think users are going to edit the configuration file manually at some point. As you can see in the following example, the structure of the JSON code is much more intuitive than the nested classes you could use in Python. If you want to store data in a dictionary in the Python file, I don't see the point of using Python for this use case.
```json
{
  "levels": {
    "default_levelup": {
      "type": "message",
      "content": "GG {user.mention}! You are now at the level {user.level}!",
      "delete_after": "5s",
    }
}```

```python
class Levels:
    class DefaultLevelup:
        type=discord.Message
        content="GG {user.mention}! You are now at the level {user.level}!"
        delete_after=datetime.timedelta(seconds=5)
```
I think using the types is the best way to do it for the Python configuration, as you don't need any extra parsing but the user needs to handle imports and to exactly know what he is doing.
If you add an extra layer of parsing, it would not be convenient for the developers, where they could just declare the type of data they are waiting for when loading the plugin or whatever, and also add a check for the type of data when the bot is loading, to prevent errors before they occurs.
If you are a new user, what type of configuration do you thing you will understand at first?

> Then it's more a question of conventions, it's also possible to apply the same type of convention using python files (ex: start every config variable with the name of the plugin, or put them in a class that have the name of the plugin ...

Just imagine there is a plugin called archive_manager, and there is a sub option called archive_duration. The key for the configuration could quickly go very long?

> or even using a unique dictionary for each plugin ! There is plenty of possible solutions). If two plugin share the same name (ie namespace), the problem is the same using json files.

As said before, what is the point of using a dictionnary where you could directly use a JSON file?

**Comment by LeiRoF** [comment]
Ok, j'ai fais une petite modification consistant à créer un fichier `core/default_config.py` qui contient les config de base du bot et auquel les config des autres plugins sont concaténés. Ce fichier n'est jamais exécuté et n'est importé que par le fichier `config.py`. De là, le fichier `conf.py` ne fait que redéfinir les variables que l'utilisateur souhaite modifier. Le fichier `setup.py` s'occupe à la fois de la concaténation des config dans `core/default_config.py` mais aussi de la création du fichier `conf.py` et l'ajout du token

Coté plugin, la config se définit via un fichier `config.py` contenant une classe ayant le même nom que le plugin:
image

Dans cette classe, le plugin peut définir tout ce qu'il souhaite (variables simples, dico, whatever)

De cette façon, la mise à jour du bot ou de ses plugins modifiera le fichier `core/default_config.py` mais pas le fichier `conf.py` ce qui rend les mises à jour transparent pour l'utilisateur (il n'aura jamais besoin de refaire son fichier de config)

**Comment by theogiraudet** [comment]
Bon je viens également donner mon avis sur le type de config, et je suis assez d'accord avec @ascpial et @Aeris1One. Pour moi, aucun intérêt de faire de la config avec un langage de programmation, et cela pour plusieurs raisons :
- La config ne contient que de la donnée, aucune logique (sinon ce n'est pas de la config mais du scripting)
- La config est censée pouvoir être éditée par quelqu'un qui n'est pas programmeur -> plus le langage pour faire cette config est puissant, plus tu as de forme pour exprimer une même config et donc plus as de chances de perdre la personne qui tente d'installer le bot
- Dans la même ligné que le point précédent : plus le langage est puissant et plus tu as de chances d'exprimer quelque chose qui risque de faire planter le programme/qui a des effets de bord. Un langage plus restreint, c'est avoir un meilleur contrôle sur ce que saisi l'utilisateur. Pour pallier à cela en restant sur du Python, ça oblige a plus documenter en indiquant ce que l'utilisateur doit taper dans la config pour ne pas tout casser, ce qui ne serait pas nécessaire sur un langage comme YAML.
- Plus compliqué de vérifier la validité de la config. Une variable qui n'existe pas et étant définie par l'utilisateur n'est pas considérée comme une erreur et une variable qui aurait dû être définie mais qui ne l'est pas ne sera détectée que au runtime et non au lancement de l'application. Ça rend la chose plus difficile à déboguer pour un utilisateur donc
-> Ce n'est pas pour rien que l'on fait de la config dans des langages spécifiques (DSL), plus le langage est restreint, moins il est risqué à utiliser

Si vous décidez tout de même de rester sur Python : ne JAMAIS faire un import comme ça de la config, ça pose de gros problèmes de sécurités (voulu par l'utilisateur, ou non). Au moins passez par le service de parsing de Python (module `ast`) pour s'assurer que l'import est safe.

**Comment by LeiRoF** [comment]
J'ai donc changé la gestion de la config pour utiliser des fichiers yaml. Il y a donc désmorais dans les plugins un fichier "config.yaml" et un fichier "setup.py" contenant une fonction "run()". Cette dernière est appelé lorsque le script `setup.py` est lancé et que l'utilisateur fait le choix de configurer le plugin en question.

Il y a donc : 
- Un fichier `core/default_config.yaml` qui contient les paramètres de configuration (avec valeurs par défaut) du bot
- Des fichiers `plugins/<plugin>/config.yaml`
- Un fichier `config.yaml` qui est créé par `setup.py` et qui override les paramètres des autres fichiers (qui doit donc être le seul modifié par l'utilisateur ... bien qu'il n'ai pas besoin vu que le script de setup propose tout ce qui faut)
- Un fichier `core/config.py` permettant de load cette config ainsi que tous les fichiers `plugins/<plugin>/config.yaml` le tout au sein du même dictionnaire `global_config`. Pour être utilisé, il faut donc faire `from core.config import global_config`

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            if choice == "":
+                error = False
+                break
+            admins = choice.replace(" ","").split(",")
+            try:
+                for admin in admins:
+                    admin = int(admin)
+                error = False
+            except:
+                print(f"{Color.Red}👑 Invalid entry. Only user ID (integers), comma and space are expected.{Color.NC}")
```
Comment (theogiraudet) [inline setup.py:136]: Je ne recommande pas spécialement l'usage des breaks qui peuvent à mon sens nuire quelque peu à la lisibilité, surtout qu'ici la ligne 128 ne sert à rien. Code alternatif sans break :
```py
error = True
while error:
    error = False
    choice = input(f"\n👑 Bot admins (User ID separated with comma. Let empty to ignore): ")
    admins = choice.replace(" ", "").split(",")
    try:
        for admin in admins:
            admin = int(admin)
    except:
        print(f"👑 Invalid entry. Only user ID (integers), comma and space are expected.")
        error = True
```

Ou code avec seulement des breaks (pour éviter de jouer et sur la condition de sortie et sur les breaks) :
```py
while True:
    choice = input(f"\n👑 Bot admins (User ID separated with comma. Let empty to ignore): ")
    admins = choice.replace(" ", "").split(",")
    try:
        for admin in admins:
            admin = int(admin)
    except:
        print(f"👑 Invalid entry. Only user ID (integers), comma and space are expected.")
        continue
    break
```

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        error = True
+        while error:
+            choice = input(f"\n{Color.Blue}🤕 Error channel (Channel ID. Let empty to ignore):{Color.NC} ")
+            if choice == "":
+                error = False
+                break
+            try:
+                channel = int(choice)
+            except:
+                print(f"{Color.Red}👑 Invalid entry. Only channel ID (integers) are expected.{Color.NC}")
```
Comment (theogiraudet) [inline setup.py:151]: Même remarque que la précédente

File: core/default_config.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+bot:
+    token: null
+    default_prefix: "!"
+    default_language: "en"
```
Comment (theogiraudet) [inline core/default_config.yaml:4]: En YML, pas besoin de mettre les guillemets qui feront partis de la valeur

File: core/config.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import yaml
+import os
+
+#################
+# GLOBAL CONFIG #
+#################
+
+global_config = {}
+
+def reload_config():
```
Comment (theogiraudet) [inline core/config.py:10]: Ajouter une petite documentation

File: start.py [hunk 5]
```diff
@@ -9,20 +9,25 @@
...
-from bot.docs import generate_docs
-from bot.config import get_config
-from utils import Gunibot, setup_logger
+from LRFutils.color import Color
+from LRFutils import log
 
 # check python version
 py_version = sys.version_info
 if py_version.major != 3 or py_version.minor < 9:
     print("Vous devez utiliser au moins Python 3.9 !", file=sys.stderr)
```
Comment (theogiraudet) [inline start.py:21]: Message à passer en anglais

**Review by theogiraudet — CHANGES_REQUESTED** [review]
Quelques petits points à revoir mais sinon ça me semble pas mal !
Je suis cependant pas fan de cela : 
> Pour être utilisé, il faut donc faire from core.config import global_config
Pour moi, un plugin ne devrait pas pouvoir avoir accès à la config des autres plugins, d'autant plus si l'accès est en édition. Je pense que la config devrait être injectée à la classe principale du plugin.

File: start.py (hunk 5 above)
Comment (LeiRoF) [reply start.py:21]: Done

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+###############
+# TOKEN CHECK #
+###############
+
+def token_set(force_set = False):
+    """Check if the token is set, if not, ask for it. Return True if the token is set, False if not."""
+
+    if config.global_config["bot"]["token"] is not None and not force_set:
+        choice = input(f"\n🔑 {Color.Blue}A token is already set. Do you want to edit it? [y/N]:{Color.NC} ")
+        if choice.lower() not in accept:
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:30]: Justement, j'utilise alternativement "in accept" ou "not in decline" selon le paramètre par défaut. De cette façon, si l'entrée ne fait partie d'aucune des deux listes, alors le comportement par défaut est sélectionné

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            with open("config.py", "w+") as conf_file:
+                conf_file.write("from core.default_config import *\n")
+
+        # Ask for the token and save it in config.py
+        with open("config.py", "a") as conf_file:
+            token = input(f"{Color.Blue}🔑 Paste your token here (let empty and press 'enter' to ignore):{Color.NC} ")
+            if token == "":
+                print(f"\n{Color.Red}❌ Setup uncomplete 🙁{Color.NC}")
+                return False
+            conf_file.write(f"\nbot.token = '{token}'\n")
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:58]: Removed.

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+def plugin_setup():
+    for plugin in os.listdir(f'plugins'):
+        if os.path.isfile(f'plugins/' + plugin + "/setup.py"):
+
+            plugin_setup = importlib.import_module(f"plugins." + plugin + ".setup")
+
+            choice = input(f"\n{Color.Blue}🔌 Do you want to configure {plugin} plugin? [Y/n]:{Color.NC} ")
+
+            if choice not in decline:
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:73]: Bien vu !

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+    config.reload_config()
+
+    install_dependencies()
+
+    token_set()
+
+    # Optional settings
+
+    choice = input(f"\n{Color.Blue}Do you want to configure optional bot settings? [Y/n]:{Color.NC} ")
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:100]: Je laisse ça à l'appréciation des futur devs qui voudront améliorer l'ergonomie ^^

Je ne suis déjà pas sûr que tous les emojis et les couleurs fonctionnent bien avec tous les terminaux donc je préfère ne pas faire trop de modification de customisation qui peuvent être potentiellement défaits plus tard (sans compter que ça n'est pas le but premier de cette PR)

File: setup.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+    if choice not in decline:
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:102]: Bien vu x2 !

File: setup.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+        # Language 
+
+        lang = "Baguete de fromage"
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:106]: Honteux de ma part...

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                print(f"{Color.Red}🌐 Invalid language.{Color.NC}")
+        if lang != "":
+            config.global_config["bot"]["default_language"] = lang.lower()
+
+        # Prefix
+
+        prefix = config.global_config["bot"]["default_prefix"]
+        choice = input(f"\n{Color.Blue}⚜️ Choose the bot command prefix? (current: {prefix}):{Color.NC} ")
+        if choice != "":
+            config.global_config["bot"]["default_prefix"] = choice
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:120]: Dans la mesure ou on ne va pas détruire tous les plugins qui n'ont pas encore le support des slash commands, oui ^^

File: setup.py (hunk 1 above)
Comment (LeiRoF) [reply setup.py:136]: Clean. J'ai appliqué ta première suggestion 👌

File: setup.py (hunk 2 above)
Comment (LeiRoF) [reply setup.py:151]: Done ✅

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+    with open("config.yaml", "w+") as conf_file:
+        yaml.dump(config.global_config, conf_file)
+
+    print(f"\n{Color.Green}✅ Setup complete!{Color.NC}")
+
+    # Start bot
+
+    choixe = input(f"\n▶️ Your config.py file is probably incomplete, which can break some features.\n\n{Color.Blue}▶️ Do you want to start the bot anyway? [Y/n]{Color.NC} ")
+    if choixe.lower() not in decline:
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:169]: Meh, c'est bien aussi d'avoir le choixe

File: core/default_config.yaml (hunk 3 above)
Comment (LeiRoF) [reply core/default_config.yaml:4]: Sûr de ça ? 🤔
Je sais que les guillemets sont optionnels, mais le fait qu'ils fassent partie de la valeur je n'en suis pas si sûr, les fichiers de traduction en comportent sans que ça mette de guillements dans les textes

Comment (theogiraudet) [reply core/default_config.yaml:4]: My bad, tu as raison

File: core/config.py (hunk 4 above)
Comment (LeiRoF) [reply core/config.py:10]: Done !

**Comment by LeiRoF** [comment]
> Pour moi, un plugin ne devrait pas pouvoir avoir accès à la config des autres plugins, d'autant plus si l'accès est en édition. Je pense que la config devrait être injectée à la classe principale du plugin.

Dans d'autres projets j'aurais été d'accord mais je doute que ça soit nécessaire ici... et je ne vois pas trop comment faire ça proprement 😅 

Le fait d'y avoir accès en lecture peut être intéressant, mais en effet, en écriture ça ne devrait pas arriver.

**Comment by theogiraudet** [comment]
> et je ne vois pas trop comment faire ça proprement 😅

Une classe s'occupe de charger les plugins (appelons-la PluginLoader ou PluginManager). Celui-ci est le seul à avoir accès à l'ensemble de la config. Au chargement du plugin (appel d'une fonction spéficique au plugin), il passe à celui-ci la config spécifique au dit plugin.

File: plugins/rss/rss.py [hunk 14]
```diff
@@ -35,7 +36,7 @@ def __init__(self, bot: Gunibot):
 
         self.embed_color = discord.Color(6017876)
         self.loop_processing = False
-        self.twitterAPI = twitter.Api(**bot.config["twitter"], tweet_mode="extended")
+        self.twitterAPI = twitter.Api(**global_config["rss"]["twitter"], tweet_mode="extended")
```
Comment (LeiRoF) [inline plugins/rss/rss.py:39]: Le dictionnaire ne semble pas s'être update dans la fonction setup mais j'avoue ne pas comprendre pourquoi :/

**Review by LeiRoF — COMMENTED** [review]
Dans plugins/rss/rss.py ligne 42:
Le dictionnaire ne semble pas s'être update dans la fonction setup (plus haut et appelée plus tôt dans l'execution) mais j'avoue ne pas comprendre pourquoi :/

**Comment by theogiraudet** [comment]
La fonction `setup` me semble non essentiel et source d'erreur dû à sa nature asynchrone. Celle-ci permet d'avoir une méthode du même nom dans chacun des cogs qui ont quant à eux un nom et un ID différent. Ainsi, lors de l’instanciation des classes, pas besoin de connaître quel classe doit être instancié vu qu'il s'agit du rôle de la fonction `setup`. Néanmoins, cette fonction introduit une redondance non nécessaire entre les différents cog et la structrure de celle-ci n'est pas très élégante : cette fonction à deux sémantiques différentes en fonction de la présence ou non d'un paramètre, elle joue donc limite le rôle de deux fonctions au lieu d'une. De plus, sa nature asynchrone est non essentiel et peut être source d'erreur. La meilleure solution pour moi et de se débarrasser de cette fonction et de passer la configuration plutôt par le constructeur de la classe du cog au lieu d'une fonction à part.
Le nouveau problème est alors d'identifier la classe qu'il faut instancier, celle-ci n'ayant pas nécessairement le même nom que le fichier qui la possède. Cela peut être fait par introspection en instanciant la classe qui hérite de `Cog`. Chaque cog devra alors stocker le bot ainsi que sa propre config dans des attributs de classe. Cela peut être évité en factorisant cette étape dans une classe mère.

Voici les nouvelles étapes que je propose
- Créer une classe `GipsyCog` qui hérite de `Cog`
- Ajouter un constructeur à cette classe qui prend le bot et la config, celui-ci initialise les deux attributs de la classe
- Faire que chaque cog actuel hérite de ce nouveau cog et délègue le traitement de la config et du bot à son parent
- Dans la fonction `dispatch` (https://github.com/Gunivers/Gipsy/blob/c4d6c3b2df3d4ca9cb3400518aa222bd620cb71a/core/config.py#L78-L87) en charge de l'instanciation des cogs et de leur inscription auprès de Discord.py, parcourir les fichiers des cogs pour trouver une classe héritant de `GipsyCog`
- Instancier la classe en question et la register auprès de Discord.py

**Comment by ascpial** [comment]
Je suis tout à fait d'accord, et ca pourra aussi permettre de charger des données liées à l'extension sans lancer le bot (par exemple vérifier la configuration)

File: bot/checks.py [hunk 15]
```diff
@@ -3,7 +3,7 @@
 
 
 def is_bot_admin(ctx: MyContext):
-    return ctx.author.id in ctx.bot.config["bot_admins"]
+    return ctx.author.id in ctx.bot.server_configs[ctx.guild.id]["admins"]
```
Comment (ZRunner) [inline bot/checks.py:6]: Le nom de cet attribut semble suggérer que la config peut être modifiée par chaque serveur, ce qui implique que n'importe qui pourrait avoir accès aux commandes admin du bot à partir de son serveur. Est-ce le cas ?

File: requirements.txt [hunk 16]
```diff
@@ -8,4 +8,5 @@ feedparser>=6.0
 python-twitter>=3.5
 async_timeout>=3.0
 argparse
-discord.py==2.0
\ No newline at end of file
+discord.py==2.0
```
Comment (ZRunner) [inline requirements.txt:11]: Je suggère d'utiliser `discord.py~=2.2.0` afin de quand même autoriser les mises à jour mineures (généralement des corrections de bug). Cf https://peps.python.org/pep-0440/#compatible-release pour la spécification syntaxique des versions de dépendances en python.

File: docs/conf.py [hunk 17]
```diff
@@ -24,7 +65,6 @@
 # The full version, including alpha/beta/rc tags
 release = "1.3"
```
Comment (ZRunner) [inline docs/conf.py:66]: A en croire le fichier start.py, on en est à la 2.0 (et probablement sur le chemin de la 3.0, mais on verra ça plus tard)

**Review by ZRunner — CHANGES_REQUESTED** [review]

File: core/default_config.yaml [hunk 18]
```diff
@@ -0,0 +1,7 @@
...
+    admins: [279568324260528128,125722240896598016] # (Default: Leirof & Z_runner, the project leaders of Gipsy)
+    error_channels: 823813751018487848                  # (Default: Hidden channel on Gunivers)
+
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply core/default_config.yaml:7]: Il s'agit de la configuration par défaut, on pourrait très bien la mettre dans le fichier python qui gère la config, mais par soucis de clarté un fichier yaml est plus indiqué. Le dossier `data` vise à contenir les données générées par les plugins, et ne contient donc aucun fichier lié au fonctionnement du bot (ce dernier peut être supprimé sans provoquer de bugs)

**Review by ascpial — APPROVED** [review]
J'ai relu les changements et testé les nouvelles fonctionnalités et tout fonctionne comme attendu !

**Review by ascpial — CHANGES_REQUESTED** [review]
Mettre à jour le README pour indiquer comment cloner les plugins de Gunivers

File: docs/conf.py (hunk 17 above)
Comment (LeiRoF) [reply docs/conf.py:66]: En effet. Le bot est sur le chemin de la 2.0 au travers de sa refonte actuelle. La version actuellement en ligne est la 1.3.
Fixé dans le prochain commit

File: requirements.txt (hunk 16 above)
Comment (LeiRoF) [reply requirements.txt:11]: Ca marche

File: setup.py [hunk 19]
```diff
@@ -1,77 +1,71 @@
 import os
 import shutil
+from tkinter import N
```
[the remark this answers is not in the document]
Comment (LeiRoF) [reply setup.py:6]: Je comprends pas ... ni le commentaire, ni d'où sort cette ligne alors que tkinter n'est utilise nulle part 🤔

[the remark this answers is not in the document]
Comment (ZRunner) [reply setup.py:6]: Généralement c'est quand un développeur entre N + tab par erreur, et que son IDE décide d'ajouter un import automatiquement. C'est pour ça qu'il faut relire ses commits :)

**Review by ZRunner — APPROVED** [review]

**Review by ascpial — APPROVED** [review]

**Comment by LeiRoF** [comment]
Le merge de cette PR est trop complexe du à l'arrivée du submodule, le merge s'est donc bricolé de cette façon :
- Branche new_structure (cette PR) renommée en beta et mise en branche par défaut
- Branche beta renommée en dev, les changements fais là bas devront être ramenés sur cette branche.
